sync/notify/bugs.py

```diff
@@ -159,34 +159,160 @@ def for_sync(sync, results):
-            summary, comment = bug_data(sync, test_results)
+            summary, comment = bug_data(sync,
+                                        test_results,
+                                        results.treeherder_url,
+                                        results.wpt_sha)
             bug_id = make_bug(summary, comment, product, component, [sync.bug])
             rv[bug_id] = [item + (link_status,) for item in test_results]
 
     return rv
 
 
-def bug_data_crash(sync, test_results):
-    summary = "New wpt crashes from PR %s" % sync.pr
-
-    comment = """The following tests have crashes in the CI runs for wpt PR %s:
-%s
-
-(getting the crash signature into these bug reports is a TODO; sorry)
-
-These updates will be on mozilla-central once bug %s lands.
-
-""" % (sync.pr,
-       detail_part(None, test_results, None, "head", False),
-       sync.bug)
-
-    comment += postscript
+class LengthCappedStringBuilder(object):
+    def __init__(self, max_length):
+        """Builder for a string that must not exceed a given length"""
+        self.max_length = max_length
+        self.data = []
+        self.current_length = 0
+
+    def append(self, other):
+        len_other = len(other)
+        if len_other + self.current_length > self.max_length:
+            return False
+        self.data.append(other)
+        self.current_length += len_other
+        return True
+
+    def has_capacity(self, chars):
+        return self.current_length + chars <= self.max_length
+
+    def get(self):
+        return "".join(self.data)
+
+
+def split_id(test_id):
+    parts = test_id.split("/")
+    last = None
+    for i, part in enumerate(parts):
+        if "#" in part or "?" in part:
+            last = i
+            break
+    if last:
+        name = "/".join(parts[i:])
+        parts = parts[:i]
+        parts.append(name)
+    return tuple(parts)
+
+
+def get_common_prefix(test_ids):
+    """Given a list of test ids, return the paths split into directory parts,
+    and the longest common prefix directory shared by all the inputs.
+
+    :param test_ids: - List of test_ids
+    :returns: ([split_name], common_prefix) The unique test_ids split on / and
+              the longest path prefix shared by all test ids (excluding filename
+              parts
+    """
+    test_ids = list(test_ids)
+    common_prefix = split_id(test_ids[0])[:-1]
+    seen_names = set()
+    split_names = []
+    for test_id in test_ids:
+        split_name = split_id(test_id)
+        if split_name in seen_names:
+            continue
+        seen_names.add(split_name)
+        split_names.append(split_name)
+        test_dir = split_name[:-1]
+        for i, (a, b) in enumerate(zip(common_prefix, test_dir)):
+            if a != b:
+                common_prefix = common_prefix[:i]
+                break
+    return split_names, common_prefix
+
+
+def make_summary(test_results, prefix, max_length=255, max_tests=3):
```

Comment:
I see the string being built little by little and stopping each it reaches the fixed limit.

It's a very iterative process. :) 

I wonder if it would have been faster to compute the full string and then reduce it under certain conditions. I have no idea, Just thinking for now.
* extract_prefix
* count the tests.
* `{prefix}` longer than maxlength bailout. ok
* create the 
     1. `{prefix} in {common_path}  [{strings of all tests}]`. This one is not addressed in the tests. I'll make a comment on tests. This is the longest possible. if it fails go to 2 or 3 or 4  or 5
     2. `{prefix} in {common_path} [{substring_series} and {test_remain_number} others]` if it fails go to 3 or 4 or 5
     3. `{prefix} in {common_path} [{tests_number} tests]` if it fails go to 4 or 5
     4. `{prefix} in {common_path}` if it fails go to 5.
     5. `{prefix}`

I even wonder if a dumb "compute all strings possible" and just select the one which fits the maxlength is not faster. Just food for thoughts.

Reply:
Computing the full string and reducing it seems appealing, but the logic still isn't linear, you can't just have a system where you add all the pieces and take one away at a time until you end up with something under the length limit because removing one part affects which other parts should remain e.g. if you have `{prefix} in {common path} [{test 1}, {test 2}, {and N others}]` then removing `{test 2}` means adjusting `N` and removing all the tests means trying an entirely different construct. So we end up with your approach which seems like it's going to be just as much logic as we already have but done in reverse. I'm not convinced it's worthwhile to spend time trying different approaches unless it's going to be a clear win.